kythe/go/serving/explore/explore.go

```diff
@@ -49,17 +57,108 @@ func (t *Tables) TypeHierarchy(ctx context.Context, req *epb.TypeHierarchyReques
 	return nil, nil
 }
 
-// Callers returns the (recursive) callers of a specified function, as a directed graph.
-// TODO: not yet implemented
+// Callers returns the callers of a specified function, as a directed graph.
 func (t *Tables) Callers(ctx context.Context, req *epb.CallersRequest) (*epb.CallersReply, error) {
-	return nil, nil
+	tickets := req.Tickets
+	if len(tickets) == 0 {
+		return nil, fmt.Errorf("missing input tickets: %v", &req)
+	}
+
+	// succMap maps nodes onto sets of successor nodes
+	succMap := make(map[string]map[string]bool)
+
+	// At the moment, this is our policy for missing data: if an input ticket has
+	// no record in the table, we don't include data for that ticket in the response.
+	// Other table access errors result in returning an error.
+	for _, ticket := range tickets {
+		succMap[ticket] = make(map[string]bool)
+		var callgraph srvpb.Callgraph
+		err := t.FunctionToCallers.Lookup(ctx, []byte(ticket), &callgraph)
+
+		if err != nil {
+			if err != table.ErrNoSuchKey {
+				return nil, fmt.Errorf("error looking up callers with ticket %q: %v", ticket, err)
+			}
+			continue // skip tickets with no mappings
+		}
+
+		if callgraph.Type != srvpb.Callgraph_CALLER {
+			return nil, fmt.Errorf("type of callgraph is not 'CALLER': %v", callgraph)
+		}
+
+		// TODO(jrtom): consider logging a warning if len(callgraph.Tickets) == 0
+		// (postprocessing should disallow this)
+		for _, predTicket := range callgraph.Tickets {
+			succMap[predTicket][ticket] = true
+		}
+	}
+
+	return &epb.CallersReply{Graph: convertSuccMapToGraph(succMap)}, nil
+}
+
+func convertSuccMapToGraph(succMap map[string]map[string]bool) *epb.Graph {
+	graph := &epb.Graph{
+		Nodes: make(map[string]*epb.GraphNode),
+	}
+
+	for ticket, succs := range succMap {
+		node := getGraphNode(graph, ticket)
+		for s, _ := range succs {
+			node.Successors = append(node.Successors, s)
+			succ := getGraphNode(graph, s)
+			succ.Predecessors = append(succ.Predecessors, ticket)
+		}
+	}
+
+	return graph
 }
 
-// Callees returns the (recursive) callees of a specified function
+func getGraphNode(graph *epb.Graph, ticket string) *epb.GraphNode {
+	node, ok := graph.Nodes[ticket]
+	if !ok {
+		node = &epb.GraphNode{}
+	}
+	return node
+}
+
+// Callees returns the callees of a specified function
 // (that is, what functions this function calls), as a directed graph.
-// TODO: not yet implemented
 func (t *Tables) Callees(ctx context.Context, req *epb.CalleesRequest) (*epb.CalleesReply, error) {
```

Comment:
Ditto comments from above.  In fact, this code is so similar, there should probably be a common helper to make the `succMap` parameterized on the table.

Reply:
Comments from above applied.  I thought about creating a helper method, but there are enough places where the code is different (text for error messages, order of keys for succMap, name of table) that I decided it wasn't worth the trouble.